### PR TITLE
Persist client state for subscriptions/presences on server

### PR DESCRIPTION
**Pull request by bolddane**
Persist client state for subscriptions & presences on the server, by means of a new Client class.  Client data is stored in persistence, keyed by the UUID and accountName passed by the client to the server, via the name_id_sync message.

/cc @zendesk/zendesk-radar

### Steps to merge
 - [ ] :+1: of the team
 - [ ] :+1: add more tests

### References
 - Jira link: https://zendesk.atlassian.net/browse/RADAR-472

### Risks
 - Low: current production client does not send name_id_sync message to server

### CODE REVIEW

File: client/client.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
+var log = require('minilog')('radar:client'),
+    Core = require('../core');
+
+function Client (name, id) {
```
Comment (vanchi-zendesk) [inline client/client.js]: accountName should be a parameter here maybe. Also, we could do the keyGet here if we have it (since it's a concern of the instance)

File: client/client.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      break;
+
+    default:
+      changed = false;
+  }
+
+  if (changed) {
+    client.lastModified = Date.now();
+    Core.Persistence.persistKey(client.key, client);
+    Core.Persistence.expire(client.key, DEFAULT_DATA_TTL);
```
Comment (vanchi-zendesk) [inline client/client.js]: These things could be put into a multi, so it happens in a single shot.

File: client/client.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+// Read client state from persistence, when it exists for a given key
+Client.dataFromPersistenceLoad = function (client) {
+  if (client.key) {
+    // When we touch the client, refresh the TTL
+    Core.Persistence.expire(client.key, DEFAULT_DATA_TTL);
+
+    // Update current client with persisted subscriptions/presences
+    Core.Persistence.readKey(client.key, function (clientOld) {
```
Comment (vanchi-zendesk) [inline client/client.js]: likewise, here too. multi removes one turn-around time.

File: client/client.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      if (clientOld) {
+        Client.clients[client.name].subscriptions = clientOld.subscriptions;
+        Client.clients[client.name].presences = clientOld.presences;
+      }
+      Client.clients[client.name].state = STATE_LOAD_DONE;
```
Comment (vanchi-zendesk) [inline client/client.js]: Where is this defined?

File: client/client.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+Client.get = function(id, name, accountName) {
+  Client.names[id] = name;
+  var client = Client._get(id, name);
+  client.key = Client._keyGet(name, accountName);
+
+  return client;
+};
+
+// For a given client id/name pair, store the associated message
+Client.dataStore = function (id, message) {
```
Comment (vanchi-zendesk) [inline client/client.js]: This can be a prototype method, so we can use this directly after we get the client, which I think is better.

```
var c = Client.get(id, name, account);

c.dataStore(message);
// similarly
c.dataLoad();
```

File: client/client.js (hunk 4 above)
Comment (vanchi-zendesk) [inline client/client.js]: Should we not replace everything here except socket id?

File: server/server.js [hunk 6]
```diff
@@ -142,24 +146,49 @@ Server.prototype._handleClientMessage = function(client, data) {
...
+
+  this._resourceMessageHandle(client, message);
+};
+
+// Initialize a client, and persist messages where required
+Server.prototype._clientDataPersist = function (id, message) {
+  // Sync the client name to the current client id
+  if (message.op == 'name_id_sync') {
+    this._clientInit(message);
+    this.clientVersion = message.client_version;
```
Comment (vanchi-zendesk) [inline server/server.js]: we probably need not store clientVersion on the server? (It will keep changing based on the client and the message). Here, can't we just use a local (stack) var instead?

File: server/server.js [hunk 7]
```diff
@@ -142,24 +146,49 @@ Server.prototype._handleClientMessage = function(client, data) {
     return;
   }
 
-  logging.info('#client.message - received', client.id, message,
-     (this.resources[message.to] ? 'exists' : 'not instantiated'),
-     (this.subs[message.to] ? 'is subscribed' : 'not subscribed')
-    );
+  if (!this._clientDataPersist(client.id, message)) {
```
Comment (vanchi-zendesk) [inline server/server.js]: Here, instead of passing the id, it's better to pass the Client (the new one) object down, so we can just do client.dataStore().

Also, this is why we probably need to rename the existing client to socket first. This way, there is no confusion between the existing cilent (which is the socket) and the new client, which you are currently adding.

File: server/server.js [hunk 8]
```diff
@@ -142,24 +146,49 @@ Server.prototype._handleClientMessage = function(client, data) {
...
     this.emit(message.op, client, message);
   }
 };
 
-// Authorize client message
-Server.prototype._messageAuthorize =  function (message, client, data) {
+// Authorize a client message
+Server.prototype._messageAuthorize =  function (message, client) {
   var rtn = Core.Auth.authorize(message, client);
   if (!rtn) {
```
Comment (vanchi-zendesk) [inline server/server.js]: suggestion: rename rtn -> is_authorized

File: tests/client.reconnect.test.js [hunk 9]
```diff
@@ -95,7 +95,7 @@ describe('When radar server restarts', function() {
   it('reestablishes presence', function(done) {
     this.timeout(4000);
     var verifySubscriptions = function() {
-      client2.presence('restore').get(function(message) {
+      client.presence('restore').get(function(message) {
```
Comment (vanchi-zendesk) [inline tests/client.reconnect.test.js]: Not sure why this change is made. Existing tests should not break.

**Comment by vanchi-zendesk** [comment]
We need tests with the name_id_sync message from the client to see if that works.

File: client/client.js (hunk 1 above)
Comment (bolddane) [inline client/client.js]: Done.

File: client/client.js (hunk 5 above)
Comment (bolddane) [inline client/client.js]: One message creates the client instance, while later messages use it, in which case the instance is located via the name/id mapping lookup.

File: client/client.js (hunk 2 above)
Comment (bolddane) [inline client/client.js]: Done.  Haven't use multi, but this is a good suggestion.

File: client/client.js (hunk 3 above)
Comment (bolddane) [inline client/client.js]: Done.

File: client/client.js (hunk 4 above)
Comment (bolddane) [inline client/client.js]: I had left in *new* code that I am not yet using, so for now I have taken it out - that is what STATE_LOAD_DONE was out of place.

I wanted a review of the current changes in the works before I go in and make the "old client" identifier <---> socket rename.  But yes, I could have done this first before any changes.

File: server/server.js (hunk 7 above)
Comment (bolddane) [inline server/server.js]: This *client.id* is the old client - i.e. socket - yes, I agree it is confusing.  I will change that tomorrow.  In server/server.js, most of the references to *client* are to the old client (i.e. socket).

File: server/server.js (hunk 6 above)
Comment (bolddane) [inline server/server.js]: I had made a note to myself about this, and then forgot to follow up on it - good catch.  

When a client sends the *name_id_sync* message, the client version is fixed for that current session, so I could use an existing hash, keyed on id (or name), or create a new hash.  I'll look into it.

File: server/server.js (hunk 8 above)
Comment (bolddane) [inline server/server.js]: rtn is a general name I use for "return value", but explicit naming is certainly better in this case.

File: tests/client.reconnect.test.js (hunk 9 above)
Comment (bolddane) [inline tests/client.reconnect.test.js]: Before I made changes, this test used to randomly break.  We can talk about it; I believe the test setup had a problem in its configuration.

File: server/server.js (hunk 6 above)
Comment (bolddane) [inline server/server.js]: I decided to store version in each client instance, as it can differ per client.

Comment (vanchi-zendesk) [inline server/server.js]: sure, that makes sense.

File: server/server.js [hunk 10]
```diff
@@ -142,31 +147,61 @@ Server.prototype._handleClientMessage = function(client, data) {
...
+  if (message.op == 'nameSync') {
+    this._clientInit(message);
+
+    socket.send({ op: 'ack', value: message && message.ack });
+    return false;
+  }
+  else {
+    var client = Client.clientGet(socket.id);
+    if (client && Semver.gte(client.version, VERSION_CLIENT_DATASTORE) &&
+        !Client.dataStore(socket.id, message)) {
```
Comment (vanchi-zendesk) [inline server/server.js]: maybe `!client.update(message)` works better?

File: server/server.js [hunk 11]
```diff
@@ -196,7 +231,13 @@ Server.prototype._persistenceSubscribe = function (name, id) {
     });
     this.subs[name] = true;
   }
-}
+};
+
+// On name_id_sync, initialize the current client
+Server.prototype._clientInit = function (initMessage) {
+  Client.dataTTLSet(this.configuration.clientDataTTL);
```
Comment (vanchi-zendesk) [inline server/server.js]: This should be set once when the server starts.

**Comment by vanchi-zendesk** [comment]
:+1: (after we use `var` the two variables.)

File: client/client.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+  Client.clients[association.name] = client;
+
+  return client;
+}
+
+// Persist subscriptions and presences
+Client.prototype.dataStore = function (message) {
+  subscriptions = this.subscriptions;
+  presences = this.presences;
```
Comment (vanchi-zendesk) [inline client/client.js]: sorry that comment went on the commit. Please use var here.
